main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification2.scala

```diff
@@ -0,0 +1,146 @@
+
+      case Result.Err((ex: TooComplexException, _, _)) =>
+        Result.Err(UnificationError.TooComplex(ex.size, loc))
+    }
+  }
+
+  /**
+    * Returns a bi-directional map from type variables to ints computed from the given list of unification equations `l`.
+    */
+  private def mkBidirectionalVarMap(l: List[(Type, Type, SourceLocation)]): Bimap[Type.Var, Int] = {
+    // Find all type variables that occur in anywhere in `l`.
+    val allVars = mutable.Set.empty[Type.Var]
+    for ((t1, t2, _) <- l) {
+      allVars ++= t1.typeVars
+      allVars ++= t2.typeVars
+    }
+
+    // Construct the forward map from type variables to ints.
+    val forward = allVars.foldLeft(Map.empty[Type.Var, Int]) {
+      case (macc, tvar) => macc + (tvar -> tvar.sym.id)
+    }
+
+    // Construct the reverse map from ints to type variables.
+    val backward = allVars.foldLeft(Map.empty[Int, Type.Var]) {
+      case (macc, tvar) => macc + (tvar.sym.id -> tvar)
+    }
+    Bimap(forward, backward)
+  }
+
+  /**
+    * Translates the given unification equation on types `p` into a unification equation on terms.
+    */
+  private def toEquation(p: (Type, Type, SourceLocation))(implicit renv: RigidityEnv, m: Bimap[Type.Var, Int]): Equation = {
+    val (tpe1, tpe2, loc) = p
+    Equation.mk(toTerm(tpe1), toTerm(tpe2), loc)
+  }
+
+  /**
+    * Returns the given type `t` as term.
+    *
+    * Uses the given bimap `m` to map type variables to term variables.
+    *
+    * The rigidity environment `renv` is used to map rigid type variables to constants and flexible type variables to term variables.
+    */
+  private def toTerm(t: Type)(implicit renv: RigidityEnv, m: Bimap[Type.Var, Int]): Term = Type.eraseTopAliases(t) match {
+    case Type.Pure => Term.True
+    case Type.Univ => Term.False
+
+    case t: Type.Var => m.getForward(t) match {
+      case None => throw InternalCompilerException(s"Unexpected unbound type variable: '$t'.", t.loc)
+      case Some(x) => renv.get(t.sym) match {
+        case Rigidity.Flexible => Term.Var(x) // A flexible variable is a real variable.
+        case Rigidity.Rigid => Term.Cst(x) // A rigid variable is a constant.
+      }
+    }
+
+    case Type.Apply(Type.Cst(TypeConstructor.Complement, _), tpe1, _) => Term.mkNot(toTerm(tpe1))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Union, _), tpe1, _), tpe2, _) => Term.mkAnd(toTerm(tpe1), toTerm(tpe2))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Intersection, _), tpe1, _), tpe2, _) => Term.mkOr(toTerm(tpe1), toTerm(tpe2))
+
+    case _ => throw InternalCompilerException(s"Unexpected type: '$t'.", t.loc)
+  }
+
+  /**
+    * Returns a Boolean substitution as a regular type substitution.
+    */
+  private def fromBoolSubst(s: FastBoolUnification.BoolSubstitution)(implicit m: Bimap[Type.Var, Int]): Substitution = {
+    Substitution(s.m.foldLeft(Map.empty[Symbol.KindedTypeVarSym, Type]) {
```

Comment:
I still have to call toMap afterwards? - so it would be an intermediate iterable of pairs.

If the keys were the same and it was flix I would map, but here fold seems appropriate.